Toyota: improve longitudinal control

cydia2020 opened the pull request:
As usual, don't have a route as I don't have access to comma servers (repaired device), but this works really well on my TSS-P Prius, best longitudinal I've seen since I started using openpilot 4 years ago.

I've done some digging after @sshane 's PCM compensation experiment and have found that the vehicle uses two acceleration values to calculate the amount of force that needs to be compensated, hence the `ACCEL_CMD_ALT` in all Toyota dbcs.

Side note, this `ACCEL_CMD_ALT` was originally named `AT_RAW` in the leaked file, which I believe is the DSU/camera's version of the un-compensated acceleration value, the vehicle only listens to `ACCEL_CMD` for actual controls, but it will still use `ACCEL_CMD_ALT` to calculate a compensatory force.

Basically, in openpilot terms, `ACCEL_CMD_ALT`, a.k.a.`AT_RAW`, is equivalent to openpilot's accel actuator (`actuators.accel`); and `ACCEL_CMD`, a.k.a. `ATACC2`, is equivalent to the min max acceleration clipped "openpilot accel actuator (`actuators.accel`) + compensatory force (`CS.pcm_neutral_force`) / vehicle weight (`self.CP.mass`)".

This PR addresses the following long-standing issues with Toyota longitudinal:
1. no more abruptions when reaching stopping state on TSS-P
3. no more slow accel request response
4. smooth SnG
5. no more cruising just below set speed when going downhill
6. no more unresponsive highway speed control

The following item(s) will need further verification:
- [x] ~~smooth engagement and accelerator depression/release transition, try to achieve stock levels of smoothness when engaging, thanks for the feedback @AlexandreSato~~ should be okay now https://github.com/commaai/openpilot/pull/30697/commits/bdc00a5177bdd8b0900232fc39a3e23cfc52d231
- [x] tuning, it seems like this tune works well on my compensated Prius, verify for other vehicles, **especially those with a pedal, verify pedal transition**
  - [x] ~~According to @CT921 , some prius V users have noticed high speed oscillations, I've seen the same thing at around 60 km/h, fix this~~ should also be okay https://github.com/commaai/openpilot/pull/30697/commits/db93dbd88f85eadb501da6c5377bc780eb1304be
- [x] ~~`ret.vEgoStopping`, verify if this value works on all vehicles, **goal: vehicle shouldn't creep**~~ works fine
- [x] ~~`ret.stoppingDecelRate = 0.8` can be optimised based on road slope to achieve a smoother stop, but is not necessary `ret.stoppingDecelRate = 0.1` is very smooth~~ 0.3 okay, no creeping
- [x] TSS 2.0, the way openpilot handles these cars is questionable, from what I've seen in Techstream, compensatory calculations will be paused when the vehicle is stopped and in standstill, but none of our TSS 2.0s go into standstill AFAIK, verify how this affects performance. TODO#1
- [x] ~~Verify if PCM behaves consistently with gas press and lever (button) press when resuming. If PCM behaves differently, check if this is something that we can control to improve responsiveness.~~ behaves the same on my 2018 prius (non-phv)

Known issue:
~~1. TSS 2.0 vehicles will creep with the SnG hack enabled (which is default openpilot behaviour)~~ Current logic seems to have fixed this, needs more testing

To-dos:
- [x] 1. ~~stop compensating on TSS 2.0 when stopping.~~ 
  - [x] test on hills, flat surface route `b4f83330b2d59205|2023-12-12--23-03-43--0`, hill route: `7069d98b0de453c0/000000e4--f987d1b8d5/0`
- [x] 2. ~~Figure out how stock Toyota pulls away so quickly behind lead, does it disable compensation?~~ Seems like PERMIT_BRAKING can solve this.

cydia2020 committed:
give carstate neutral force

Co-Authored-By: Shane Smiskol <shane@smiskol.com>

cydia2020 committed:
apply offset to pcm_accel_cmd

Co-Authored-By: Shane Smiskol <shane@smiskol.com>

cydia2020 committed:
simplify Toyota long tunes

cydia2020 committed:
give PCM raw acceleration

cydia2020 committed:
improve starting accel

cydia2020 committed:
seems to be okay on mine?

cydia2020 committed:
don't compensate on NO_STOP_TIMER_CAR when stopping

adeebshihadeh commented:
Nice! I'm gonna convert to draft until your TODOs are done.

cydia2020 committed:
Revert "seems to be okay on mine?"

This reverts commit a3b6f9de80d44acc93a92e56499c7eab8fe752b1.

cydia2020 committed:
0.3 works okay

foodbandlt commented:
Route on 2022 Camry Hybrid TSS2, has some interesting bookmarks: b4f83330b2d59205|2023-12-12--23-03-43--0

Will edit this comment with more once I get some better examples, but it performed REALLY well and felt more like how I would control the throttle.  Little to no oscillations on highway in experimental mode, too!  Thanks for the awesome work!

sshane commented on selfdrive/car/toyota/interface.py:
```diff
@@ -253,23 +253,14 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long, docs):
     # to a negative value, so it won't matter.
     ret.minEnableSpeed = -1. if (stop_and_go or ret.enableGasInterceptor) else MIN_ACC_SPEED
 
+    # on stock Toyota this is -2.5
+    ret.stopAccel = -2.5
+
```
this is -0.4 on TSS2

0312birdzhang commented on selfdrive/car/toyota/carstate.py:
```diff
@@ -146,6 +146,7 @@ def update(self, cp, cp_cam):
       ret.cruiseState.standstill = self.pcm_acc_status == 7
     ret.cruiseState.enabled = bool(cp.vl["PCM_CRUISE"]["CRUISE_ACTIVE"])
     ret.cruiseState.nonAdaptive = cp.vl["PCM_CRUISE"]["CRUISE_STATE"] in (1, 2, 3, 4, 5, 6)
+    self.pcm_neutral_force = cp.vl["PCM_CRUISE"]["NEUTRAL_FORCE"]
```
is this `self` a typo?

cydia2020 replied:
Nope

cydia2020 committed:
set permit braking bit according to accel

cydia2020 committed:
Merge remote-tracking branch 'upstream/master' into toyota_pcm_compensation

cydia2020 committed:
Fix line too long

AlexandreSato commented:
Three consecutive bookmarks mean uphill, two downhill and one flat surface. Posted in redundancy in the Pull Request comment "Toyota Improve longitudinal control" and discord thread to gain traction.

On commit "Fix line too long" 62cc41a:
```075b133b6181e058|2023-12-16--12-10-44```
```075b133b6181e058|2023-12-16--12-24-38```
```075b133b6181e058|2023-12-16--12-56-50```
```075b133b6181e058|2023-12-16--13-04-06```

2023-12-16 
> [!TIP]
> Positives on my corolla TSS2 hybrid: No more abrupt accelerations, more responsive on uphill exits;
> [!CAUTION]
> Negative, it brakes unnecessarily when engaging.

ntegan1 commented:
feels like a big improvement, tssp corolla pedal transition too.
62cc41a5b7343851092805f78e17e520437e4c48
a152dad3c3368320|2023-12-22--06-33-26
a152dad3c3368320|2023-12-21--17-05-10
a152dad3c3368320|2023-12-21--12-26-29
a152dad3c3368320|2023-12-21--12-21-54
a152dad3c3368320|2023-12-21--12-06-36
a152dad3c3368320|2023-12-21--07-48-27
a152dad3c3368320|2023-12-21--07-16-58
a152dad3c3368320|2023-12-20--17-27-32
a152dad3c3368320|2023-12-20--17-04-07
a152dad3c3368320|2023-12-20--07-53-50

AlexandreSato commented:
As my previous stock routes were lost because I forgot to preserve, I'm posting new ones, better and this time preserved and public:
075b133b6181e058|2023-12-24--13-41-53
075b133b6181e058|2023-12-24--13-36-36--0

cydia2020 committed:
Merge remote-tracking branch 'upstream/master' into toyota_pcm_compensation

cydia2020 committed:
prevent braking when releasing accelerator

cydia2020 committed:
add back deadzone

cydia2020 committed:
fix

cydia2020 committed:
adhd

AlexandreSato commented:
on 5c789cb "adhd" It's no longer speeding up or slowing down.
On highway:     075b133b6181e058|2023-12-27--12-17-49
city: 075b133b6181e058|2023-12-27--16-38-17

cydia2020 committed:
well lol

AlexandreSato commented:
On [77d3108](https://github.com/commaai/openpilot/pull/30697/commits/77d31084f3fd17d95352af61ceb76b959eef7281)
Brake when engage fixed! flawless absolutely no jerk or harsh just a bit of abrupt brake when entering in stopped state machine.
Some routes:
075b133b6181e058|2023-12-28--14-12-50
075b133b6181e058|2023-12-28--15-27-40
075b133b6181e058|2023-12-28--17-51-12

cydia2020 committed:
Merge remote-tracking branch 'upstream/master' into toyota_pcm_compensation

cydia2020 committed:
very low stopping decel

Seems to stop smoothly without creep

cydia2020 committed:
remove `should_compensate` on SnG hack cars

cydia2020 committed:
forgot this

cydia2020 committed:
here as well

cydia2020 committed:
Fix checks error

sshane commented on selfdrive/car/toyota/toyotacan.py:
```diff
@@ -33,18 +33,19 @@ def create_lta_steer_command(packer, steer_control_type, steer_angle, steer_req,
     "CANCEL_REQ": pcm_cancel,
     "ALLOW_LONG_PRESS": 1,
     "ACC_CUT_IN": fcw_alert,  # only shown when ACC enabled
+    "ACCEL_CMD_ALT": accel_raw,
```
We should test if this actually provides any improvement. On the TSS2 car I've checked (Camry Hybrid), it's almost always identical.

sshane commented on selfdrive/car/toyota/toyotacan.py:
```diff
@@ -33,18 +33,19 @@ def create_lta_steer_command(packer, steer_control_type, steer_angle, steer_req,
     "DISTANCE": 0,
     "MINI_CAR": lead,
-    "PERMIT_BRAKING": 1,
+    "PERMIT_BRAKING": 1 if accel < 0.3 else 0,
```
I'd be careful with this, if you're on hills, the point at which you need to brake is much higher, so this adds opportunities for late braking. I tested messing with this signal, but could never get anything safe in all situations.

I recall Toyota keeping this bit at 1 behind leads, and acceleration feels fine there. It's possible that this is used for comfort when cruising without leads (to engine brake only)

cydia2020 replied:
This is fine I think, I actually have this on my own fork for quite a while now, usually when going uphill the pcm will just stop accelerating, this should be okay, however if it's too low, say 0, it will create oscillations when going downhill which we dont want

sshane commented on selfdrive/car/toyota/interface.py:
```diff
@@ -256,23 +256,17 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long, docs):
     # to a negative value, so it won't matter.
     ret.minEnableSpeed = -1. if (stop_and_go or ret.enableGasInterceptor) else MIN_ACC_SPEED
 
+    # on stock Toyota this is -2.5
+    ret.stopAccel = -2.5
```
Have a route for this? Which cars does this apply to?

cydia2020 replied:
Prius, this is what I've grabbed from my scanner

sshane commented on selfdrive/car/toyota/interface.py:
```diff
@@ -256,23 +256,17 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long, docs):
-      tune.kiBP = [0., 35.]
-      tune.kpV = [3.6, 2.4, 1.5]
-      tune.kiV = [0.54, 0.36]
+    tune.kpBP = [0.,]
+    tune.kiBP = [0., 3.]
+    tune.kpV = [1.]
+    tune.kiV = [.3, 1.]
```
I worry this PR changes too many things, making it harder to validate and merge. If there is one or two concrete fixes, that's easier to reason about.

sshane commented on selfdrive/car/toyota/carcontroller.py:
```diff
@@ -114,7 +114,15 @@ def update(self, CC, CS, now_nanos):
       interceptor_gas_cmd = 0.
-    pcm_accel_cmd = clip(actuators.accel, self.params.ACCEL_MIN, self.params.ACCEL_MAX)
+
+    if CC.longActive:
+      accel_offset = CS.pcm_neutral_force / self.CP.mass
+    else:
+      accel_offset = 0.
```
I don't know if we fully understand this signal yet, I just labeled my best guess looking at some data. Is there anything in techinfo, or the dealer software about this?

cydia2020 replied:
only thing in the repair manual that I can find relating to feedback force (which I believe this is) of any kind is Cruise Request Driving Force Feedback Status, which I've also mentioned in my opendbc issue a few weeks ago (https://github.com/commaai/opendbc/issues/987), it has a few states but I don't exactly know which CAN ID this might be on, it might relate to how this is calculated in the stock system

sshane commented on selfdrive/controls/controlsd.py:
```diff
@@ -586,7 +586,8 @@ def state_control(self, CS):
     CC.latActive = self.active and not CS.steerFaultTemporary and not CS.steerFaultPermanent and \
                    (not standstill or self.joystick_mode)
-    CC.longActive = self.enabled and not self.events.contains(ET.OVERRIDE_LONGITUDINAL) and self.CP.openpilotLongitudinalControl
+    CC.longActive = self.enabled and (not self.events.contains(ET.OVERRIDE_LONGITUDINAL) or self.CP.carName == 'toyota') \
+                   and self.CP.openpilotLongitudinalControl
```
What is this for?

cydia2020 replied:
reverted, tried to smooth gas presses out with this hack but doesn't seem to work all the time, and created other issues.

cydia2020 committed:
Revert "prevent braking when releasing accelerator"

This reverts commit 22f77bde9234de524e0d0344d99f429a02f0345b.

cydia2020 committed:
revert this

cydia2020 committed:
Revert "remove `should_compensate` on SnG hack cars"

This reverts commit 85279b9fde399b11cac21fec473908111f957f6c.

cydia2020 committed:
match ACCEL_CMD_ALT with compensate if leadVisible

cydia2020 committed:
Merge remote-tracking branch 'upstream/master' into toyota_pcm_compensation

cydia2020 committed:
Revert "Fix checks error"

This reverts commit 8bda9cb6476353f577c77cd305238c28ac8f8126.

cydia2020 committed:
revert these

AlexandreSato commented:
On 'revert these' 759f0c8:
075b133b6181e058|2024-01-05--12-03-55 (Public and preserved)
Fixed hard braking when entering Stopped State. I believe creep happened here: 075b133b6181e058|2024-01-05--12-03-55--2

AlexandreSato commented:
In fact, what is happening is that STOP_DISTANCE is not consistent (see 075b133b6181e058|2024-01-06--16-23-56--4). However, the longitudinal control by openpilot is now perfectly the same as stock Toyota, very good!

cydia2020 committed:
send raw accel -2.5 when stopping

cydia2020 committed:
don't allow negative calculations at first

Co-Authored-By: Alexandre Nobuharu Sato <66435071+alexandresato@users.noreply.github.com>

cydia2020 committed:
Merge remote-tracking branch 'upstream/master' into toyota_pcm_compensation

cydia2020 committed:
use a standalone condition for stopping in toyotacan.py

cydia2020 committed:
Revert "use a standalone condition for stopping in toyotacan.py"

This reverts commit f829fc11bef471b6018abc717b3471617861c2c1.

cydia2020 committed:
Revert "send raw accel -2.5 when stopping"

This reverts commit 6aa88102de771a887fc74c3a35434a3b5e8c91ae.

cydia2020 committed:
Revert "don't allow negative calculations at first"

This reverts commit bdc00a5177bdd8b0900232fc39a3e23cfc52d231.

cydia2020 committed:
handle accel_raw in carcontroller.py

cydia2020 committed:
retry: prohibit negative compensation when first enabling

thanks @AlexandreSato for the idea

Co-Authored-By: Alexandre Nobuharu Sato <66435071+AlexandreSato@users.noreply.github.com>

cydia2020 committed:
remove extra arg

cydia2020 committed:
remove trailing whitespace

cydia2020 committed:
simpler logic

cydia2020 committed:
add some comments

cydia2020 committed:
more comments

cydia2020 committed:
Merge remote-tracking branch 'upstream/master' into toyota_pcm_compensation

cydia2020 committed:
remove extra pos arg

cydia2020 committed:
make this global

cydia2020 committed:
reduce deadzone to as small as possible

cydia2020 committed:
revert this for TSS2.0

cydia2020 committed:
set raw accel to -2.5 when lead is stopped

thanks @Smartype for the initial implementation of the lead stopped logic

Co-Authored-By: Eric Day <smartype@gmail.com>

cydia2020 committed:
remove this

AlexandreSato commented:
on: https://github.com/commaai/openpilot/commit/4bc10ba3309924dd54191073deb2ddf555e6b69e
best longitudinal control seen in years. Only improvements without any regression on my corolla TSS2. Creep definitely solved! In my opinion it's good to go. Public and preserved route:
https://connect.comma.ai/075b133b6181e058/1706014680345/1706016181471
`075b133b6181e058|2024-01-23--09-58-02`

cydia2020 committed:
bump cereal

cydia2020 committed:
bump opendbc

cydia2020 committed:
Merge remote-tracking branch 'upstream/master' into toyota_pcm_compensation

ntegan1 commented:
commit aa895f7282583cb092356cd8fe99c177fe7d79ee

a152dad3c3368320|2024-01-23--19-42-30 public 9 segment route
similar experience with tssp corolla + pedal

cydia2020 committed:
use a lower threshold for beginning neg calcs

cydia2020 committed:
Merge remote-tracking branch 'upstream/master' into toyota_pcm_compensation

cydia2020 committed:
fix misspelling

cydia2020 committed:
.15 is a bit slow sometimes

cydia2020 commented:
Also we might be able to fix #29613 once this is merged, as this does introduce a lead stopped condition, and that can be used to fix the standstill entrance behaviour

```python
      if CS.out.vEgo < self.CP.vEgoStopping and lead_vehicle_stopped and self.CP.carFingerprint not in NO_STOP_TIMER_CAR:
        self.standstill_req = True
      else:
        self.standstill_req = False
```

cydia2020 committed:
disable compensatory calculations when stopped

jyoung8607 commented:
> Also we might be able to fix #29613 once this is merged, as this does introduce a lead stopped condition, and that can be used to fix the standstill entrance behaviour

May I ask what you're trying to accomplish by considering the lead car's stopped condition? It's abstraction-breaking and I'm not sure what it gains you from reading this PR and other associated history. Does it improve the final stop position? Smoothness? Does it prevent creep on resume press? If there's some material improvement, then surely the same improvement is desired for stopping at a red light with no lead car visible.

cydia2020 commented:
> > Also we might be able to fix #29613 once this is merged, as this does introduce a lead stopped condition, and that can be used to fix the standstill entrance behaviour
> 
> May I ask what you're trying to accomplish by considering the lead car's stopped condition? It's abstraction-breaking and I'm not sure what it gains you from reading this PR and other associated history. Does it improve the final stop position? Smoothness? Does it prevent creep on resume press? If there's some material improvement, then surely the same improvement is desired for stopping at a red light with no lead car visible.

Yea it might not be needed tbh, anecdotally and from my own testing it seems to have helped with creeping

jyoung8607 commented:
> Yea it might not be needed tbh, anecdotally and from my own testing it seems to have helped with creeping

To expand on what I was saying before: in an E2E world, making policy decisions based on lead vehicle speed really belongs in higher levels of openpilot. However, if the Toyota CAN signal `ACC_CONTROL.LEAD_VEHICLE_STOPPED` is a real thing and the vehicle powertrain/chassis is verified to react to it in a way that we understand and find useful, the Toyota car port can absolutely (ab)use that signal to obtain desired stopping/starting performance. It might, for example, be used to let go of the brakes earlier and prepare for departure without actually applying engine throttle yet. I'm just guessing what its effect might be, I would have to observe the stock system working.

AlexandreSato commented:
On [3b89906](https://github.com/commaai/openpilot/pull/30697/commits/3b89906cecc5a7475fe142428e67ed28410a3fe5)

Very good experience!
```
075b133b6181e058|2024-01-24--07-33-19
```

cydia2020 committed:
lower stoppingDecelRate for TSS 2+

Thanks @AlexandreSato for testing

Co-Authored-By: Alexandre Nobuharu Sato <66435071+alexandresato@users.noreply.github.com>

cydia2020 committed:
Revert "disable compensatory calculations when stopped"

This reverts commit ff9a14313871a08d99af06d854a4bd81cfceb175.

cydia2020 committed:
Merge remote-tracking branch 'upstream/master' into toyota_pcm_compensation

AlexandreSato commented:
good stops at https://github.com/commaai/openpilot/pull/30697/commits/ef15ad3d66bbf4fb9a3b9f3acdd9c9b4c3a9094b
```
075b133b6181e058|2024-01-26--10-39-08
```

cydia2020 committed:
Revert "Revert "disable compensatory calculations when stopped""

This reverts commit 1194d4b6b93d5f1446c2ee14444d3fff7d09238c.

cydia2020 committed:
Merge branch 'master' into toyota_pcm_compensation

sshane commented on selfdrive/car/toyota/toyotacan.py:
```diff
@@ -33,18 +33,20 @@ def create_lta_steer_command(packer, steer_control_type, steer_angle, steer_req,
     "ALLOW_LONG_PRESS": 1,
     "ACC_CUT_IN": fcw_alert,  # only shown when ACC enabled
+    "ACCEL_CMD_ALT":  accel_raw,  # raw accel command, pcm uses this to calculate a compensatory force
+    "LEAD_VEHICLE_STOPPED": lead_stopped,
```
This bit is never set on TSS2 Camry Hybrid

sshane commented:
The stock system mostly keeps PERMIT_BRAKING at 1 when it's behind a lead, even during takeoff, so I think we can achieve better sng performance without touching it (I'd really like to avoid using it too since it's not easy to get a binary brake control bit like this to smoothly and safely work, especially on hills)

https://github.com/commaai/openpilot/assets/25857203/ef8ec657-1a72-493f-b97d-30e57512e802

pd0wm commented:
On the stock system of the 2021 Rav4 Prime it seems like `ACCEL_CMD` is always 0, and only `ACCEL_CMD_ALT` is used. (`20ba9ade056a8c7b|2021-02-08--21-57-35`). Is this the first car that shows this behavior?

cydia2020 commented:
> The stock system mostly keeps PERMIT_BRAKING at 1 when it's behind a lead, even during takeoff, so I think we can achieve better sng performance without touching it (I'd really like to avoid using it too since it's not easy to get a binary brake control bit like this to smoothly and safely work in all scenarios, especially on hills)
> 
> https://github.com/commaai/openpilot/assets/25857203/74878c55-ae4b-4efe-a670-eb3ceaf1cd87

Yup agreed, also I've realised that setting this bit to 0 at lower speeds does lead to some jerkiness when openpilot is letting go of the brakes, I'll get rid of it

AlexandreSato commented:
Good behavior no issues on ![image](https://github.com/commaai/openpilot/assets/66435071/62fa3eac-a8c1-49c6-ba3a-edecfd70ac83)
```
075b133b6181e058|2024-01-30--11-22-14
```
https://connect.comma.ai/075b133b6181e058/1706624534176/1706625197269
Eco and normal mode, public and preserved

```
075b133b6181e058|2024-01-30--11-01-25
```
preserved

cydia2020 committed:
Merge branch 'master' into toyota_pcm_compensation

cydia2020 committed:
reset cereal

cydia2020 committed:
remove LVSTP from toyotacan and revert permit braking to stock

AlexandreSato commented:
Very good at 'remove LVSTP from toyotacan and revert permit braking to stock' a8c3f6a13610996f1d34f98d4fbbeaddb4b123a8
```
075b133b6181e058|2024-02-02--21-10-12
```
```
075b133b6181e058|2024-02-02--19-11-04
```

cydia2020 committed:
Merge branch 'master' into toyota_pcm_compensation

foodbandlt commented:
Hey!  So I've jumped around stock Toyota long, OP long on `master` and OP long on this PR branch and this branch gets much closer to how stock Toyota long reacts with gas and brake (which is better than OP long).  It's still much smoother than OP long on `master`, but seems to have gained some strange throttle oscillations that I haven't found the cause of yet.

Here's a preserved route on `5fee813` with bookmarks: `b4f83330b2d59205|2024-02-08--22-44-39--0` (https://connect.comma.ai/b4f83330b2d59205/1707432279382/1707434769358)
* Seg 9 has a very well executed hard stop
* Segs 15-17 have starts, stops, and oscillations when stopping.  Perhaps while starting too, I don't remember.
* Seg 23 has oscillations when approaching stopped vehicle
* Seg 24 has a nice hard stop
* Seg 30 has oscillations at highway speeds when the lead slows down drastically
* Seg 39 I believe had oscillations when approaching slower lead, but this is still by far the best braking performance I've had on any branch or fork! 

Before the permit braking bit was removed, engaging while stopped behind a lead was broken and started idling forward with no intention of stopping.  I'll test it again on this commit shortly and edit my comment.

cydia2020 commented:
> Hey! So I've jumped around stock Toyota long, OP long on `master` and OP long on this PR branch and this branch gets much closer to how stock Toyota long reacts with gas and brake (which is better than OP long). It's still much smoother than OP long on `master`, but seems to have gained some strange throttle oscillations that I haven't found the cause of yet.
> 
> Here's a preserved route on `5fee813` with bookmarks: `b4f83330b2d59205|2024-02-08--22-44-39--0` (https://connect.comma.ai/b4f83330b2d59205/1707432279382/1707434769358)
> 
> * Seg 9 has a very well executed hard stop
> * Segs 15-17 have starts, stops, and oscillations when stopping.  Perhaps while starting too, I don't remember.
> * Seg 23 has oscillations when approaching stopped vehicle
> * Seg 24 has a nice hard stop
> * Seg 30 has oscillations at highway speeds when the lead slows down drastically
> * Seg 39 I believe had oscillations when approaching slower lead, but this is still by far the best braking performance I've had on any branch or fork!
> 
> Before the permit braking bit was removed, engaging while stopped behind a lead was broken and started idling forward with no intention of stopping. I'll test it again on this commit shortly and edit my comment.

Throttle oscillation is usually tuning related, I've noticed it on mine as well when decelerating over a distance downhill. Tuning is something that I'll need to look into a bit more, as of now I can still identify some imperfections with stock MPC, hopefully tuning can reduce those.

cydia2020 committed:
ramp up comp threshold at higher speed

cydia2020 committed:
Merge remote-tracking branch 'upstream/master' into toyota_pcm_compensation

cydia2020 committed:
reduce low speed agressiveness

cydia2020 committed:
add some comments for the new changes

cydia2020 committed:
can get away with this

cydia2020 committed:
reduce low speed oscillation

cydia2020 committed:
Merge branch 'master' into toyota_pcm_compensation

AlexandreSato commented:
Very good at xxxx (2024-02-14)
Public and preserved test routes
```
075b133b6181e058|2024-02-14--11-43-21
```
https://connect.comma.ai/075b133b6181e058/1707921801231/1707922285843

cydia2020 committed:
Merge branch 'master' into toyota_pcm_compensation

cydia2020 committed:
Merge branch 'master' into toyota_pcm_compensation

cydia2020 committed:
do we really need to differentiate?

eFiniLan commented on selfdrive/car/toyota/carcontroller.py:
```diff
@@ -136,11 +163,11 @@ def update(self, CC, CS, now_nanos):
         can_sends.append(toyotacan.create_acc_cancel_command(self.packer))
       elif self.CP.openpilotLongitudinalControl:
-        can_sends.append(toyotacan.create_accel_command(self.packer, pcm_accel_cmd, pcm_cancel_cmd, self.standstill_req, lead, CS.acc_type, fcw_alert,
+        can_sends.append(toyotacan.create_accel_command(self.packer, pcm_accel_cmd, actuators.accel, pcm_cancel_cmd, self.standstill_req, lead, CS.acc_type, fcw_alert,
```
Is this supposed to be accel_raw? It looks like a merge issue.

cydia2020 replied:
Thanks a lot for pointing this out! Fixed. <3

cydia2020 committed:
Merge branch 'master' into toyota_pcm_compensation

cydia2020 committed:
fix this

cydia2020 commented:
Route: 7069d98b0de453c0/000000e4--f987d1b8d5/0

Updated this PR for the new longitudinal API, also updated the tuning for TSS-P vehicles, this tune seems to produce minimal oscillations for all conditions on my 2018 Prius, route is included above for your convenience.

cydia2020 committed:
general formatting

cydia2020 committed:
fix line too long

sshane commented:
We've moved the car interfacing code to our [`opendbc`](https://github.com/commaai/opendbc) repository, which is now the new home for car ports and fingerprints. Please re-open your pull request against [`opendbc`](https://github.com/commaai/opendbc) at your convenience by using this command below. This will transform all changes under `selfdrive/car/` to `opendbc_repo/opendbc/car/`. Make sure you have initialized submodules beforehand and have checked out the latest `opendbc` commit.

```bash
PR_NUMBER=33045
curl -L https://github.com/commaai/openpilot/pull/$PR_NUMBER.patch | sed -e 's/selfdrive\/car/opendbc_repo\/opendbc\/car/g' | git apply -v --reject
```

Simply replace the PR number with your own. Once done, add the files, fix any conflicts, and open a new PR. Alternatively, you may start a new PR from scratch if that is easier for you.
